File: app/src/main/java/com/islam/basepropject/project_base/utils/others/ViewModelFactory.java
package com.islam.basepropject.project_base.utils.others;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.islam.basepropject.ui.ExFetchData.Fragment1;
import com.islam.basepropject.ui.ExViewPager.FragmentTab1;
import com.islam.basepropject.ui.Fragment2;
import com.islam.basepropject.ui.ExPaging.ExViewModel;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private static ViewModelFactory INSTANCE;
    ;

    public static ViewModelFactory getInstance() {
        if (INSTANCE == null) {
            synchronized (ViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ViewModelFactory();
                }
                return INSTANCE;
            }
        }
        return INSTANCE;
    }

    private ViewModelFactory() {}

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass.isAssignableFrom(Fragment1.ViewModel.class)) {
            return (T) new Fragment1.ViewModel();
        }
        if (modelClass.isAssignableFrom(Fragment2.ViewModel.class)) {
            return (T) new Fragment2.ViewModel();
        }
        if (modelClass.isAssignableFrom(ExViewModel.class)) {
            return (T) new ExViewModel();
        }

        if (modelClass.isAssignableFrom(FragmentTab1.ViewModel.class)) {
            return (T) new FragmentTab1.ViewModel();
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }

}

File: app/src/main/java/com/islam/basepropject/project_base/utils/others/ViewModelFactory.kt
package com.islam.basepropject.project_base.utils.others

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import com.islam.basepropject.ui.ExFetchData.Fragment1
import com.islam.basepropject.ui.ExRecyclerView.Fragment3
import com.islam.basepropject.ui.ExViewPager.FragmentTab1
import com.islam.basepropject.ui.Fragment2
import com.islam.basepropject.ui.ExPaging.ExViewModel

class ViewModelFactory private constructor() : ViewModelProvider.NewInstanceFactory() {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {

        if (modelClass.isAssignableFrom(Fragment1.ViewModel::class.java)) {
            return Fragment1.ViewModel() as T
        }
        if (modelClass.isAssignableFrom(Fragment2.ViewModel::class.java)) {
            return Fragment2.ViewModel() as T
        }
        if (modelClass.isAssignableFrom(ExViewModel::class.java)) {
            return ExViewModel() as T
        }

        if (modelClass.isAssignableFrom(FragmentTab1.ViewModel::class.java)) {
            return FragmentTab1.ViewModel() as T
        }

        if (modelClass.isAssignableFrom(Fragment3.ViewModel::class.java)) {
            return Fragment3.ViewModel() as T
        }
        throw IllegalArgumentException("Unknown ViewModel class: " + modelClass.name)
    }

    companion object {
        private var INSTANCE: ViewModelFactory? = null

        val instance: ViewModelFactory
            get() {
                if (INSTANCE == null) {
                    synchronized(ViewModelFactory::class.java) {
                        if (INSTANCE == null) {
                            INSTANCE = ViewModelFactory()
                        }
                        return INSTANCE as ViewModelFactory
                    }
                }
                return INSTANCE as ViewModelFactory
            }
    }

}
